csaf-validation/src/main/kotlin/io/github/csaf/sbom/validation/tests/Tests.kt

```diff
@@ -0,0 +1,289 @@
+fun List<Test>.test(doc: Csaf): ValidationResult {
+    return this.map { it.test(doc) }.merge()
+}
+
+/**
+ * Implementation of
+ * [Test 6.1.1](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#611-missing-definition-of-product-id).
+ */
+object Test611MissingDefinitionOfProductID : Test {
+    override fun test(doc: Csaf): ValidationResult {
+        val definitions = doc.gatherProducts().map { it.product_id }
+        val references = doc.gatherProductReferences()
+
+        val notDefined = references.subtract(definitions.toSet())
+
+        return if (notDefined.isEmpty()) {
+            ValidationSuccessful
+        } else {
+            ValidationFailed(
+                listOf("The following IDs are not defined: ${notDefined.joinToString(",")}")
+            )
+        }
+    }
+}
+
+/**
+ * Implementation of
+ * [Test 6.1.2](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#612-multiple-definition-of-product-id).
+ */
+object Test612MultipleDefinitionOfProductID : Test {
+    override fun test(doc: Csaf): ValidationResult {
+        val definitions = doc.gatherProducts().map { it.product_id }
+
+        val duplicates = definitions.duplicates()
+
+        return if (duplicates.isEmpty()) {
+            ValidationSuccessful
+        } else {
+            ValidationFailed(
+                listOf("The following IDs are duplicate: ${duplicates.keys.joinToString(",")}")
+            )
+        }
+    }
+}
+
+private fun <T> List<T>.duplicates(): Map<T, Int> {
+    return groupingBy { it }.eachCount().filter { it.value > 1 }
+}
+
+/**
+ * Implementation of
+ * [Test 6.1.3](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#613-circular-definition-of-product-id).
+ */
+object Test613CircularDefinitionOfProductID : Test {
+    override fun test(doc: Csaf): ValidationResult {
+        val circles = mutableSetOf<String>()
+
+        for (relationship in doc.product_tree?.relationships ?: listOf()) {
+            val definedId = relationship.full_product_name.product_id
+            val notAllowed =
+                listOf(relationship.product_reference, relationship.relates_to_product_reference)
+            if (definedId in notAllowed) {
+                circles += definedId
+            }
+        }
+
+        return if (circles.isEmpty()) {
+            ValidationSuccessful
+        } else {
+            ValidationFailed(
+                listOf("The following IDs are defined in circles: ${circles.joinToString(",")}")
+            )
+        }
+    }
+}
+
+/**
+ * Implementation of
+ * [Test 6.1.4](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#614-missing-definition-of-product-group-id).
+ */
+object Test614MissingDefinitionOfProductGroupID : Test {
+    override fun test(doc: Csaf): ValidationResult {
+        val definitions = doc.gatherProductGroups().map { it.group_id }
+        val references = doc.gatherProductGroupReferences()
```

Comment:
The spec only lists `/vulnerabilities[]/remediations[]/group_ids` and `/vulnerabilities[]/threats[]/group_ids` as relevant paths. However, we consider `(/vulnerabilities[])+/remediations[]/group_ids`. Is this intended, mistake in the code (unnecessary) or missing in the spec?

Reply:
No, we only consider these two. There are not nested vulnerabilities, it's extremely hard to read in the code. I think I will try to rewrite all of these functions, hopefully I can still test them the same then.

Reply:
I know that this doesn't exist based on the schema. Yet, if I understand the Extensions correctly, I think they could be called in a cycle and there's no control which would enforce a certain pattern based on the path here.

Reply:
Got it, I reworked this now to remove these extra functions.